### PR TITLE
impl(generator): add LoggingDecorator for REST

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236)

### CODE REVIEW

File: generator/integration_tests/golden/internal/golden_thing_admin_rest_logging_decorator.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  StatusOr<google::test::admin::database::v1::ListBackupOperationsResponse> ListBackupOperations(
+      google::cloud::rest_internal::RestContext& rest_context,
+      google::test::admin::database::v1::ListBackupOperationsRequest const& request) override;
+
+ private:
+  std::shared_ptr<GoldenThingAdminRestStub> child_;
+  TracingOptions tracing_options_;
+  std::set<std::string> components_;
+};  // GoldenThingAdminRestLogging
```
Comment (coryan) [inline generator/integration_tests/golden/internal/golden_thing_admin_rest_logging_decorator.h:115]: This closing comment seems useless.  And we only generated it for the `*Logging` decorators, can we remove it?

File: generator/integration_tests/tests/golden_kitchen_sink_rest_logging_decorator_test.cc [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+namespace cloud {
+namespace golden_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::testing::Contains;
+using ::testing::HasSubstr;
+using ::testing::Return;
+
+class LoggingDecoratorRestTest : public ::testing::Test {
```
Comment (coryan) [inline generator/integration_tests/tests/golden_kitchen_sink_rest_logging_decorator_test.cc:34]: This is another class where the fixture makes the code harder to read than just defining `mock_` and `log_` locally in each test.

File: generator/integration_tests/tests/golden_thing_admin_rest_logging_decorator_test.cc [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+namespace golden_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::testing::Contains;
+using ::testing::HasSubstr;
+using ::testing::Return;
+using ::testing::Unused;
+
+class LoggingDecoratorRestTest : public ::testing::Test {
```
Comment (coryan) [inline generator/integration_tests/tests/golden_thing_admin_rest_logging_decorator_test.cc:34]: Same comment about the fixture as in the other test.

File: generator/internal/logging_decorator_rest_generator.cc [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+// If you make any local changes, they will be lost.
+// source: $proto_file_name$
+
+)""");
+
+  // includes
+  CcLocalIncludes({vars("logging_rest_header_path"),
+                   "google/cloud/internal/log_wrapper.h",
+                   "google/cloud/status_or.h"});
+  CcSystemIncludes({vars("proto_header_path"), "memory"});
```
Comment (coryan) [inline generator/internal/logging_decorator_rest_generator.cc:113]: I think the `#include` for `vars("proto_header_path")` belongs in the generated `.h` file or nowhere.  If you are trying to use IWYU, then it is needed in the header file.  OTOH `vars("stub_rest_header_path")` already provides it, so I am not even sure it is needed there.

**Review by coryan — APPROVED** [review]
I think we can clean up the generated code a little bit, but it is good enough as-is.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236)* status: 0 of 13 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[`generator/integration_tests/golden/internal/golden_thing_admin_rest_logging_decorator.h` line 115 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236#-NGctZCc5ZTU7cQeCPeD-r2-115:-NGrnWqH8fFgo7X6RFqb:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9672b1fbf2e33996b4304ad5917611bb404fff9b/generator/integration_tests/golden/internal/golden_thing_admin_rest_logging_decorator.h#L115)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This closing comment seems useless.  And we only generated it for the `*Logging` decorators, can we remove it?
</blockquote></details>

Done.
___
*[`generator/integration_tests/tests/golden_kitchen_sink_rest_logging_decorator_test.cc` line 34 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236#-NGd39d-4R43WbiyGda4-r2-34:-NGrr_eM0Q5mtqy57xmG:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9672b1fbf2e33996b4304ad5917611bb404fff9b/generator/integration_tests/tests/golden_kitchen_sink_rest_logging_decorator_test.cc#L34)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This is another class where the fixture makes the code harder to read than just defining `mock_` and `log_` locally in each test.
</blockquote></details>

Done.
___
*[`generator/integration_tests/tests/golden_thing_admin_rest_logging_decorator_test.cc` line 34 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236#-NGd39d-4R43WbiyGda5-r2-34:-NGrr_1aCrwiEfFRWcP8:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9672b1fbf2e33996b4304ad5917611bb404fff9b/generator/integration_tests/tests/golden_thing_admin_rest_logging_decorator_test.cc#L34)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Same comment about the fixture as in the other test.
</blockquote></details>

Done.
___
*[`generator/internal/logging_decorator_rest_generator.cc` line 113 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10236#-NGctZCc5ZTU7cQeCPeI-r2-113:-NGrnZRZ0DWIAgkRwaJY:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9672b1fbf2e33996b4304ad5917611bb404fff9b/generator/internal/logging_decorator_rest_generator.cc#L113)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think the `#include` for `vars("proto_header_path")` belongs in the generated `.h` file or nowhere.  If you are trying to use IWYU, then it is needed in the header file.  OTOH `vars("stub_rest_header_path")` already provides it, so I am not even sure it is needed there.
</blockquote></details>

Done.

**Pull request merged**